### PR TITLE
Open project file via file association

### LINKED ISSUE #149: Open project file instantaneously, skipping welcome menu.
Skip startup menu when project file (preferably `.yafc`) is specified in command line arguments.

# Additional Features
 - Add startup menu shortcut to recent project file automatically.

### PR DESCRIPTION
**Pull request by shpaass**
A successor of https://github.com/shpaass/yafc-ce/pull/152. I've made this PR so I can discuss some aspects.
> If only one command-line option is given, assume that it is a project file and attempt to open it.
Closes #149

### CODE REVIEW

**Comment by shpaass** [comment]
The problem though is that it will eat CI hours that we're alotted to. I need to look into how to disable CI checks for drafts.

**Comment by shpaass** [comment]
Alright, I've pushed the changes to the CI. Let's see if they work.

**Comment by shpaass** [comment]
The run does not start. CI changes seem to be working.

**Comment by shpaass** [comment]
QA:
✅ Loaded a vanilla and the main pY-AE project from CLI.
✅ Loaded the same two projects with the Open With menu on Windows.
✅ Loaded the same two projects with a double-click after choosing to always open `.yafc` files with Yafc.
✅ Saved a vanilla project after loading it from CLI and changing it.
❌ Crashed after changing and trying to save the pY-AE project. 
The crash happened after adding/removing rows or sheets, and *only* when loading the project through CLI.

The error log is below. Sharing it just in case. I'll take a look at it in the coming days.
```
Yafc.Model.ProductionTable.modules must not return null.

   at Yafc.Model.PropertySerializer`2.getter(TOwner owner) in .\Yafc.Model\Serialization\PropertySerializers.cs:line 61
   at Yafc.Model.ReadOnlyReferenceSerializer`2.SerializeToJson(TOwner owner, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\PropertySerializers.cs:line 102
   at Yafc.Model.SerializationMap`1.SerializeToJson(T value, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\SerializationMap.cs:line 229
   at Yafc.Model.ModelObjectSerializer`1.WriteToJson(Utf8JsonWriter writer, T value) in .\Yafc.Model\Serialization\ValueSerializers.cs:line 113
   at Yafc.Model.ValuePropertySerializer`2.SerializeToJson(TOwner owner, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\PropertySerializers.cs:line 74
   at Yafc.Model.SerializationMap`1.SerializeToJson(T value, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\SerializationMap.cs:line 229
   at Yafc.Model.ModelObjectSerializer`1.WriteToJson(Utf8JsonWriter writer, T value) in .\Yafc.Model\Serialization\ValueSerializers.cs:line 113
   at Yafc.Model.CollectionSerializer`3.SerializeToJson(TOwner owner, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\PropertySerializers.cs:line 171
   at Yafc.Model.SerializationMap`1.SerializeToJson(T value, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\SerializationMap.cs:line 229
   at Yafc.Model.SerializationMap`1.SpecificSerializationMap.SerializeToJson(Object target, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\SerializationMap.cs:line 86
   at Yafc.Model.ReadOnlyReferenceSerializer`2.SerializeToJson(TOwner owner, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\PropertySerializers.cs:line 110
   at Yafc.Model.SerializationMap`1.SerializeToJson(T value, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\SerializationMap.cs:line 229
   at Yafc.Model.ModelObjectSerializer`1.WriteToJson(Utf8JsonWriter writer, T value) in .\Yafc.Model\Serialization\ValueSerializers.cs:line 113
   at Yafc.Model.CollectionSerializer`3.SerializeToJson(TOwner owner, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\PropertySerializers.cs:line 171
   at Yafc.Model.SerializationMap`1.SerializeToJson(T value, Utf8JsonWriter writer) in .\Yafc.Model\Serialization\SerializationMap.cs:line 229
   at Yafc.Model.Project.Save(String fileName) in .\Yafc.Model\Model\Project.cs:line 112
   at Yafc.MainScreen.SaveProject() in .\Yafc\Windows\MainScreen.cs:line 649
   at Yafc.MainScreen.KeyDown(SDL_Keysym key) in .\Yafc\Windows\MainScreen.cs:line 574
   at Yafc.UI.InputSystem.KeyDown(SDL_Keysym key) in .\Yafc.UI\Core\InputSystem.cs:line 83
   at Yafc.UI.Ui.ProcessEvents() in .\Yafc.UI\Core\Ui.cs:line 110
```

**Comment by shpaass** [comment]
Reminded by @Dorus, I agree that the frustration of setting up two full paths is much bigger than the frustration of crashing during the load if one had wrong pre-fills, and then changing the paths as necessary. Therefore, pre-fills are advised.

- [ ] Pre-fill data and mods paths with something reasonable if this project has not been opened by Yafc before.

**Comment by shpaass** [comment]
Regarding the crash, it crashes in writing multiple properties called `variants`. Not sure what it is yet.

**Comment by shpaass** [comment]
It seems that it might be just my pY-AE project that fails to save because it fails even without CLI involved. 
At the same time, a new pY-AE project is saved successfully with both CLI and vanilla way of opening.

**Comment by shpaass** [comment]
Aaaaaallllriiiiight. Master fails to save my project. Bisection time.

**Comment by shpaass** [comment]
```
$ git bisect good 0.8.1
[5 bisections later]
7d55f43e58e48a8d031c4980ea237097a659ab70 is the first bad commit
commit 7d55f43e58e48a8d031c4980ea237097a659ab70
Author: Dale McCoy <21223975+DaleStan@users.noreply.github.com>
Date:   Wed Aug 21 00:43:46 2024 -0400

    test(fixed-production): Add tests for the fixed amounts feature.

 .../Model/ProductionTableContentTests.cs           | 148 ++++++++++
 .../Model/ProductionTableContentTests.lua          | 318 +++++++++++++++++++++
 Yafc.Model.Tests/Yafc.Model.Tests.csproj           |   4 +
 Yafc.Model/Model/ProductionTable.cs                |   2 +-
 .../ProductionTable/ProductionTableView.cs         |   8 +-
 5 files changed, 473 insertions(+), 7 deletions(-)
 create mode 100644 Yafc.Model.Tests/Model/ProductionTableContentTests.cs
 create mode 100644 Yafc.Model.Tests/Model/ProductionTableContentTests.lua
```

I'll make a ticket.

**Comment by shpaass** [comment]
@DaleStan this revert https://github.com/shpaass/yafc-ce/pull/244/commits/eeec967cd098670a192700855d4bee1918c34454 fixes the crash https://github.com/shpaass/yafc-ce/issues/245, but I wanted to check with you if it keeps your changes operational.

By the way, my local tests fail to run for some reason. Says `Test run aborted`. I wonder if you have the same output.

**Comment by shpaass** [comment]
The preliminary code is ready, but I'll need to do the following QA tomorrow:
```
QA:
All actions are done for both vanilla and modded.
    // the usual
        Add a row
        Add a page
        Save the project, close Yafc
        Load the project.
    
    // regression test for ProjectDefinition constructor
        Load and close a project
        Load the Welcome Screen
        Load the project from the list of recent ones
    
    // regression test for not-null-forgiving ProjectDefinition
        Make a new project on the Welcome Screen
        Make a new project in the CLI
        
    // net production CLI test
        Make a vanilla project
        Check that Kovarex displays has no looping item.
```

**Comment by shpaass** [comment]
The QA mentioned above is done.

**Comment by shpaass** [comment]
One thing to check on your side guys, is that this PR doesn't bork your list of recent projects.

**Comment by shpaass** [comment]
Also, please let me know if the commits are too granular or too big.

**Comment by shpaass** [comment]
I heavily advise to review this PR through the Files Changed tab, commit by commit.
Otherwise it might be hard to see why something was done this way.

File: Yafc.Model/Serialization/PropertySerializers.cs [hunk 1]
```diff
@@ -157,13 +160,9 @@ public override void DeserializeFromJson(TOwner owner, ref Utf8JsonReader reader
             base.DeserializeFromJson(owner, ref reader, context);
         }
 
-        public override void SerializeToUndoBuilder(TOwner owner, UndoSnapshotBuilder builder) {
-            builder.WriteManagedReference(getter(owner) ?? throw new InvalidOperationException($"Cannot serialize a null value for {property.DeclaringType}.{propertyName} to the undo snapshot."));
-        }
+        public override void SerializeToUndoBuilder(TOwner owner, UndoSnapshotBuilder builder) => builder.WriteManagedReference(getter(owner) ?? throw new InvalidOperationException($"Cannot serialize a null value for {property.DeclaringType}.{propertyName} to the undo snapshot."));
```
[the remark this answers is not in the document]
Comment (shpaass) [reply Yafc.Model/Serialization/PropertySerializers.cs:163]: Today I learned that Code Cleanup doesn't even try to make lines not-long. Thank you for catching this!

**Comment by shpaass** [comment]
Alright, the default diff has both my changes and the merged PRs, so to see the actual changes, I used
```
$ git range-diff master..646cc master..09e20
 1:  dfe72db !  1:  a383158 feat(cli): skip the Welcome Screen when given only a path in CLI
    @@ Yafc/Utils/Preferences.cs: namespace Yafc {

      ## changelog.txt ##
     @@ changelog.txt: Date:
    -         - Add a setting to force software rendering if hardware rendering does not work on your system.
              - Switch tabs using Ctrl+PgUp / Ctrl+PgDown
              - Reorder tabs using Ctrl+Shift+PgUp / Ctrl+Shift+PgDown
    +         - Add 'Produce it as a spent fuel' recipe selection.
     +        - You can now open projects from a Windows context-menu. If the project was not opened before, then Yafc uses
     +          the launch-settings from the most-recently opened project.
          Bugfixes:
 2:  7d16a37 =  2:  86ec8e6 docs(cli): add the path-only case to the help
 3:  3cc720f =  3:  dc0826d chore: improve docs wording
 4:  b3606fa =  4:  b8b22a7 chore(git): ignore user-specific launch settings for Debug
 5:  d418a29 =  5:  873270d style: run Code Cleanup
 6:  3f2310d !  6:  6a95dda refactor(serialization): refactor getters to increase readability
    @@ Yafc.Model/Serialization/PropertySerializers.cs: namespace Yafc.Model {
     +                }
     +            }
     +
    -+            return _getter(owner);
    ++            return result;
     +        }


 7:  5f5164d =  7:  bc85e9f style: cosmetic changes
 8:  7f636db =  8:  5e27b16 refactor(Preferences.cs): use constructor for ProjectDefinition
 9:  7f3aca2 =  9:  2e0a86c refactor(cli:parse): rename options to projectDefinition
10:  ef0f238 = 10:  ffbae8a refactor: make ProjectDefinition parameters not-null-forgiving
11:  363b3e6 = 11:  7735aa8 fix: add net production as a CLI option
12:  68bc8b5 = 12:  e4f0107 refactor: reorder the members of ProjectDefinition constructor
13:  27f2466 = 13:  e9e5659 docs(WelcomeScreen): add a TODO about ProjectDefinition usage
14:  5454b7b = 14:  87143fd fix: allow creating new projects from CLI
15:  78e6d58 = 15:  6384913 docs: add a suggestion to keep line length below 120
16:  cc11f62 = 16:  ab095d8 docs(cli): add summary to CommandLineParser
17:  2220544 = 17:  fe93e13 refactor(cli): rename Cli Parse into ParseArgs
18:  dd26c62 = 18:  87bab5a refactor(Preferences): change order of parameters in AddProject()
19:  af4be2a = 19:  5990cc0 refactor(MainScreen): In SaveProjectAs(), rename path to projectPath
20:  da7fd52 = 20:  67fc3b3 style: apply Code Cleanup to MainScreen.cs
21:  9082b0c = 21:  18930e8 docs(Preferences): add a TODO about deserialization (253)
22:  78a50ee = 22:  6860e82 refactor(Preferences): In AddProject() rename path to projectPath
23:  646cced = 23:  3e09617 docs(Preferences): add docs to ProjectDefinition
 -:  ------- > 24:  09e20a5 style(PropertySerializers.cs): manually fix line length

```

File: Yafc.Model/Serialization/PropertySerializers.cs [hunk 2]
```diff
@@ -57,38 +53,45 @@ internal abstract class PropertySerializer<TOwner, TPropertyType>(PropertyInfo p
...
+                    return default;
+                }
+                else {
+                    throw new InvalidOperationException($"{property.DeclaringType}.{propertyName} must not return null.");
+                }
+            }
 
-        public override void DeserializeFromJson(TOwner owner, ref Utf8JsonReader reader, DeserializationContext context) {
-            setter(owner, ValueSerializer.ReadFromJson(ref reader, context, owner));
+            return _getter(owner);
```
[the remark this answers is not in the document]
Comment (shpaass) [reply Yafc.Model/Serialization/PropertySerializers.cs]: ...I'll just go crawl under a rug.
xD
Addressed in https://github.com/shpaass/yafc-ce/pull/244/commits/6a95ddad9155364e8e2a6ffe64f4cce7a0e9aca6

File: Yafc.Model/Serialization/PropertySerializers.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (shpaass) [reply Yafc.Model/Serialization/PropertySerializers.cs]: Addressed in https://github.com/shpaass/yafc-ce/pull/244/commits/09e20a5b90f32fa452872fa7eafc649b78d156a3

**Review by DaleStan — APPROVED** [review]

**Pull request merged**